packages/apollo-server-defer/README.md

```diff
@@ -0,0 +1,173 @@
+
+- Figuring out how to chop up your query into different pieces and wrapping them with multiple query components
+- Sharing data between queries may be problematic (e.g. secondary queries might depend on some data returned from the first query)
+- Breaking up queries like this just doesn't feel like the smooth GraphQL experience we want 😢
+
+#### A Better Way
+
+_@defer_ provides a way for us to mark fields in our schema to _not_ wait on, because they might be:
+
+- _Expensive to load_. This includes private data that is not cached (like user progress), or information that requires more computation on the backend (like calculating price quotes on Airbnb)
+- _Not on the critical path for interactivity._ This includes loading the comments section of a blog post, or the number of claps received.
+- _Expensive to send._ Even if the field may resolve quickly (ready in memory), users might still choose to defer it if the cost of transport is too expensive.
+
+#### Use Cases
+
+```graphql
+type Video {
+  # public data that can be cached
+  id: ID!
+  title: String
+  description: String
+  # progress is private to the user, and significantly more
+  # expensive to retrieve
+  progress: Progress
+  comments: [Comment]
+}
+
+type Progress {
+  percent: Float
+  position: Float
+}
+
+type Comment {
+  userId: ID!
+  text: String
+}
+
+# Example Query using @defer
+query {
+  video(id: $id) {
+    title: String
+    progress @defer {
+      percent
+      position
+    }
+    comments @defer {
+      userId
+      text
+    }
+  }
+}
+```
+
+Instead of holding the GraphQL response and page rendering until the entire query is resolved, @defer tells Apollo Server to return a partial query response if the deferred fields are not ready yet. In the example above, we see how it may be used to remove `progress` and `comments` from the intial query response. Apollo Server would then take care of resolving the rest of the deferred fields in the background, and stream them to the client when they are ready.
+
+```
+// Initial Response
+{
+  "asset": {
+    "title": "Westworld",
+  }
+}
+
+// Stream in deferred fields as patches
+{
+  "path": ["video", "progress"],
+  "data": {
+    "percent": 20,
+    "position": 20
+  }
+}
+{
+  "path": ["video", "comments"],
+  "data": [{
+    "userId": "123",
+    "text": "..."
+  }]
+}
+```
+
+#### Who needs this the most?
+
+- Media companies, banks, other data-heavy enterprise applications.
+
+#### Why is @defer an ideal solution?
+
+- Query remains super easy to read 👍
+- Stays true to Apollo's declarative approach to data loading - everything is encapsulated in a single query component 😎
+- Greatly reduces boilerplate code to manage multiple requests, a huge value-add for users that use both Apollo Server AND Client together.
+
+### Implementation Details
+
+#### Patch Data Format:
+
+- Proposed format for `ExecutionPatchResult`
+  ```graphql
+  {
+    path: [string]!
```

Comment:
Does this mean I can always tell apart a patch from a non-patch by checking if it has a `path`? that's nice!

Reply:
Correct!